libopenage/gamestate/map.cpp

```diff
@@ -0,0 +1,89 @@
+// Copyright 2024-2024 the openage authors. See copying.md for legal info.
+
+#include "map.h"
+
+#include <nyan/nyan.h>
+
+#include "gamestate/api/terrain.h"
+#include "gamestate/terrain.h"
+#include "gamestate/terrain_chunk.h"
+#include "pathfinding/cost_field.h"
+#include "pathfinding/grid.h"
+#include "pathfinding/pathfinder.h"
+#include "pathfinding/sector.h"
+
+
+namespace openage::gamestate {
+Map::Map(const std::shared_ptr<Terrain> &terrain) :
+	terrain{terrain},
+	pathfinder{std::make_shared<path::Pathfinder>()},
+	grid_lookup{} {
+	// Get grid types
+	// TODO: This should probably not be dependent on the existing tiles, but
+	//       on all defined nyan PathType objects
+	size_t grid_idx = 0;
+	auto chunk_size = this->terrain->get_chunk(0)->get_size();
+	auto side_length = std::max(chunk_size[0], chunk_size[0]);
+	util::Vector2s grid_size{this->terrain->get_row_size(), this->terrain->get_column_size()};
+	for (const auto &chunk : this->terrain->get_chunks()) {
+		for (const auto &tile : chunk->get_tiles()) {
+			if (tile.terrain != std::nullopt) {
+				auto path_costs = api::APITerrain::get_path_costs(*tile.terrain);
+
+				for (const auto &path_cost : path_costs) {
+					if (not this->grid_lookup.contains(path_cost.first)) {
+						auto grid = std::make_shared<path::Grid>(grid_idx, grid_size, side_length);
+						this->pathfinder->add_grid(grid);
+
+						this->grid_lookup.emplace(path_cost.first, grid_idx);
+						grid_idx += 1;
+					}
+				}
+			}
+		}
+	}
+
+	// Set path costs
+	for (size_t chunk_idx = 0; chunk_idx < this->terrain->get_chunks().size(); ++chunk_idx) {
+		auto chunk_terrain = this->terrain->get_chunk(chunk_idx);
+		for (size_t tile_idx = 0; tile_idx < chunk_terrain->get_tiles().size(); ++tile_idx) {
+			auto tile = chunk_terrain->get_tile(tile_idx);
+			if (tile.terrain != std::nullopt) {
+				auto path_costs = api::APITerrain::get_path_costs(*tile.terrain);
+
```

Comment:
Can you invert the if condition and make it an early `continue` and then reduce the nesting level of the code after:
```c++
if (tile.terrain == std::nulopt) {
    continue;
}
auto path_costs = api::APITerrain::get_path_costs(*tile.terrain);
...
```

Reply:
true, although I just removed that if-branch altogether :D